<a href="https://colab.research.google.com/github/AvishekPritom/Pandas-learning-path/blob/main/Pandas_03_How_do_I_select_a_subset_of_a_DataFrame.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

 Titanic Dataset Loading....

In [ ]:
titanic = pd.read_csv('/content/titanic.csv')

In [ ]:
titanic.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


#How do I select specific columns from a DataFrame?

I’m interested in the age of the Titanic passengers.

In [ ]:
ages = titanic['Age']

In [ ]:
ages

,Age
0,22.0
1,38.0
2,26.0
3,35.0
4,35.0
...,...
886,27.0
887,19.0
888,NaN
889,26.0


Note

> To select a single column, use square brackets [] with the column name of the column of interest.



We can verify this by checking the type of the output:

In [ ]:
type(titanic['Age'])

pandas.core.series.Series

I’m interested in the age and sex of the Titanic passengers.

In [ ]:
agessex = titanic[['Age','Sex']]

Note

>To select multiple columns, use a list of column names within the selection brackets [].The inner square brackets define a Python list with column names, whereas the outer square brackets are used to select the data from a pandas DataFrame as seen in the previous example.



In [ ]:
titanic[['Age','Sex']].shape

(891, 2)

The selection returned a DataFrame with 891 rows and 2 columns. Remember, a DataFrame is 2-dimensional with both a row and column dimension.

#How do I filter specific rows from a DataFrame?

I’m interested in the passengers older than 35 years.

In [ ]:
above_35 = titanic[titanic['Age']>35]

In [ ]:
above_35.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,C103,S
13,14,0,3,"Andersson, Mr. Anders Johan",male,39.0,1,5,347082,31.2750,NaN,S
15,16,1,2,"Hewlett, Mrs. (Mary D Kingcome)",female,55.0,0,0,248706,16.0000,NaN,S


Note

>To select rows based on a conditional expression, use a condition inside the selection brackets [].
The condition inside the selection brackets titanic["Age"] > 35 checks for which rows the Age column has a value larger than 35:
These outer brackets mean:
“Select/filter rows from the DataFrame.”


In [ ]:
titanic['Age'] > 35

,Age
0,False
1,True
2,False
3,False
4,False
...,...
886,False
887,False
888,False
889,False


Note

> The difference between the titanic['Age'] > 35 and titanic[['Age','Sex']].shape is when titanic[] is selecting the whole dataframe and titanic['Age'] > 35 first selecting the attribute and then checks the boolean condition



I’m interested in the Titanic passengers from cabin class 2 and 3.

In [ ]:
class_23 = titanic[titanic["Pclass"].isin([2, 3])]

Note

> Similar to the conditional expression, the isin() conditional function returns a True for each row the values are in the provided list. To filter the rows based on such a function, use the conditional function inside the selection brackets []. In this case, the condition inside the selection brackets titanic["Pclass"].isin([2, 3]) checks for which rows the Pclass column is either 2 or 3.



In [ ]:
aclass_23 = titanic[(titanic['Pclass'] == 2) | (titanic['Pclass'] == 3)]

In [ ]:
aclass_23.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S


Note

> When combining multiple conditional statements, each condition must be surrounded by parentheses (). Moreover, you can not use or/and but need to use the or operator | and the and operator &.



I want to work with passenger data for which the age is known.

In [ ]:
age_no_null = titanic[titanic['Age'].notna()]

In [ ]:
age_no_null.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


Note

>The notna() conditional function returns a True for each row the values are not a Null value. As such, this can be combined with the selection brackets [] to filter the data table.



You might wonder what actually changed, as the first 5 lines are still the same values. One way to verify is to check if the shape has changed:

In [ ]:
age_no_null.shape

(714, 12)

#How do I select specific rows and columns from a DataFrame?

I’m interested in the names of the passengers older than 35 years

In [ ]:
adult_name = titanic.loc[titanic['Age'] > 35 , 'Name']

Note

> In this case, a subset of both rows and columns is made in one go and just using selection brackets [] is not sufficient anymore. The loc/iloc operators are required in front of the selection brackets []. When using loc/iloc, the part before the comma is the rows you want, and the part after the comma is the columns you want to select.

>When using column names, row labels or a condition expression, use the loc operator in front of the selection brackets []. For both the part before and after the comma, you can use a single label, a list of labels, a slice of labels, a conditional expression or a colon. Using a colon specifies you want to select all rows or columns.



I am interested in rows 10 till 25 and columns 3 to 5.

In [ ]:
titanic.iloc[10:24 , 3:4]

,Name
10,"Sandstrom, Miss. Marguerite Rut"
11,"Bonnell, Miss. Elizabeth"
12,"Saundercock, Mr. William Henry"
13,"Andersson, Mr. Anders Johan"
14,"Vestrom, Miss. Hulda Amanda Adolfina"
15,"Hewlett, Mrs. (Mary D Kingcome)"
16,"Rice, Master. Eugene"
17,"Williams, Mr. Charles Eugene"
18,"Vander Planke, Mrs. Julius (Emelia Maria Vande..."
19,"Masselmani, Mrs. Fatima"


Note
>Again, a subset of both rows and columns is made in one go and just using selection brackets [] is not sufficient anymore. When specifically interested in certain rows and/or columns based on their position in the table, use the iloc operator in front of the selection brackets [].

>When selecting specific rows and/or columns with loc or iloc, new values can be assigned to the selected data. For example, to assign the name anonymous to the first 3 elements of the fourth column:

In [ ]:
titanic.iloc[0:3, 3] = "anonymous"
titanic.iloc[:5, 3]

,Name
0,anonymous
1,anonymous
2,anonymous
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)"
4,"Allen, Mr. William Henry"


**Remember**


*  When selecting subsets of data, square brackets [] are used.
*  Inside these square brackets, you can use a single column/row label, a list of column/row labels, a slice of labels, a conditional expression or a colon.
*   Use loc for label-based selection (using row/column names).
*   Use iloc for position-based selection (using table positions).
*   You can assign new values to a selection based on loc/iloc.




